pkg/controller/deployment_inplace.go

```diff
@@ -0,0 +1,584 @@
+	klog.V(3).Infof("New machine set %s has %d available machines.", newMachineSet.Name, newMachineSet.Status.AvailableReplicas)
+	maxUnavailable := MaxUnavailable(*deployment)
+
+	minAvailable := deployment.Spec.Replicas - maxUnavailable
+	newISUnavailableMachineCount := newMachineSet.Spec.Replicas - newMachineSet.Status.AvailableReplicas
+	oldISsMachinesUndergoingUpdate, err := dc.getMachinesUndergoingUpdate(oldMachineSets)
+	if err != nil {
+		return false, err
+	}
+
+	klog.V(3).Infof("allMachinesCount:%d,  minAvailable:%d,  newISUnavailableMachineCount:%d,  oldISsMachineInUpdateProcess:%d", allMachinesCount, minAvailable, newISUnavailableMachineCount, oldISsMachinesUndergoingUpdate)
+
+	// maxUpdatePossible is calculated as the total number of machines (allMachinesCount)
+	// minus the minimum number of machines that must remain available (minAvailable),
+	// minus the number of machines in the new instance set that are currently unavailable (newISUnavailableMachineCount),
+	// minus the number of machines in the old instance sets that are undergoing updates (oldISsMachinesUndergoingUpdate).
+	// here unavailable machines of old machine sets are not considered as first we want to check if we can select machines for update from old machine sets
+	// after fulfilling all the constraints.
+	maxUpdatePossible := allMachinesCount - minAvailable - newISUnavailableMachineCount - oldISsMachinesUndergoingUpdate
+	if maxUpdatePossible <= 0 {
+		klog.V(3).Infof("no machines can be selected for update from old machine sets")
+		return false, nil
+	}
+
+	// prepare machines from old machine sets for update, need to check maxUnavailable to ensure we can select machines for update.
+	numOfMachinesSelectedForUpdate, err := dc.selectNumOfMachineForUpdate(ctx, allMachineSets, oldMachineSets, newMachineSet, deployment, oldISsMachinesUndergoingUpdate)
+	if err != nil {
+		return false, err
+	}
+
+	return numOfMachinesSelectedForUpdate > 0, nil
+}
+
+func (dc *controller) selectNumOfMachineForUpdate(ctx context.Context, allMachineSets []*v1alpha1.MachineSet, oldMachineSets []*v1alpha1.MachineSet, newMachineSet *v1alpha1.MachineSet, deployment *v1alpha1.MachineDeployment, oldISsMachinesUndergoingUpdate int32) (int32, error) {
+	maxUnavailable := MaxUnavailable(*deployment)
+
+	// Check if we can pick machines from old ISes for updating to new IS.
+	minAvailable := deployment.Spec.Replicas - maxUnavailable
+
+	// Find the number of available machines.
+	availableMachineCount := GetAvailableReplicaCountForMachineSets(allMachineSets) - oldISsMachinesUndergoingUpdate
+	if availableMachineCount <= minAvailable {
+		// Cannot pick for updating.
+		return 0, nil
+	}
+
+	sort.Sort(MachineSetsByCreationTimestamp(oldMachineSets))
+
+	totalSelectedForUpdate := int32(0)
+	maxSelectableForUpdate := min(availableMachineCount-minAvailable, max(deployment.Spec.Replicas-newMachineSet.Spec.Replicas, 0))
+	for _, targetIS := range oldMachineSets {
```

Comment:
can you rename `targetIS` with `targetMachineSet`? Same in other places in this function.